google/datalab/contrib/bigquery/operators/bq_execute_operator.py

```diff
@@ -0,0 +1,38 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+# or implied. See the License for the specific language governing permissions and limitations under
+# the License.
+
+import google
+import google.datalab.bigquery as bq
+import pickle
+from airflow.models import BaseOperator
+from airflow.utils.decorators import apply_defaults
+from google.datalab.contrib.pipeline._pipeline import Pipeline
+
+
+class ExecuteOperator(BaseOperator):
+
+  @apply_defaults
+  def __init__(self, query, parameters, table, mode, py_context_str, *args, **kwargs):
+    super(ExecuteOperator, self).__init__(*args, **kwargs)
+    self._query = query
+    self._table = table
+    self._mode = mode
+    self._parameters = parameters
+    self._py_context_str = py_context_str
```

Comment:
I introduced this yesterday, and don't have this working end-to-end yet (seeing a few issues when unit-testing the round-tripping of a pickled object through a text representation because mock objects cannot pickled apparently). 

To answer your question though, bigquery.Query.execute() takes the context object as a parameter, so we need this in our operator:
https://github.com/googledatalab/pydatalab/blob/bdaac31bbfa36e3e21de743c1d6d8fb7f6b3b8ab/google/datalab/bigquery/_query.py#L322

@nikhilk : I wondered about this though, since it's decidedly ugly. Am I on the wrong track in believing that this context object is necessary?

Reply:
Done. _construct_context_for_args() was the helper in Datalab that got this, so I moved that to the Context class and am using that everywhere now. No more pickling.